### ISSUE TITLE
contenful images to local folders

### ISSUE THREAD

**Issue by ergunpp** [gatsbyjs/gatsby#10587]
When building a site with data/images from contentful, is there a way to copy images to local folder and generate links to local copies?

**Comment by sidharthachatterjee** [comment]
Hey @ergunpp 

From what I understand, the `gatsby-source-contentful` doesn't exactly support copying images locally and generating links to the those. 

You could however use `createRemoteFileNode` in `gatsby-source-filesystem` to download files from contentful and add them to your site's GraphQL schema. 

Check out the example at https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-source-filesystem#readme

**Comment by polarathene** [comment]
@ergunpp I wanted this as well, I've got something going that will download any ContentfulAsset to the local filesystem and provide a localFilename node that you can use for fluid images with gatsby-image, those get built/copied to the public/static folder. The images will link to these instead(you have to use a slightly different qraphql query and update code accordingly. 

I'll look into doing a PR with what I have so far.

@sidharthachatterjee Thanks for the tip with `createRemoteFileNode` and link to example, I referenced the wordpress code that it links to from there and managed to get something similar with contentful. I'm not 100% if I've done it correctly, sometimes it seems to error out when I run `gatsby develop` that it's unable to find/query the fieldname I assigned, I'm guessing because it's async with the download and I need to have the build wait until the images are downloaded first(and their nodes updated with the custom field for querying).

I could create a PR for others if someone who understands it better can review/advise.

**Comment by jonniebigodes** [comment]
@polarathene i've done some work on this issue over the weekend. But if you have something already, please do share a repository and i don't mind looking into it and proceed from there, if you don't mind.

**Comment by polarathene** [comment]
@sidharthachatterjee Hi, yes I've figured out the issue :) I was hacking away at the node_module I had in my project and it was converted into JS which didn't have async/await syntax(generators instead), although I could still use async/await just fine, just using await within the async sourceNodes method(which was converted to a generator and yields replacing awaits), it wasn't working. Original source was in a nested dir, copy/paste built with that just fine :)

I have it working now and without any cache/race condition I was running into earlier.I was utilizing existingNodes to add the new local file property, but this doesn't exist if there isn't a cache.. I have copied the same logic and placed after the calls to the two custom createNode methods contentful uses and issue resolved.

The wordpress plugin that I referenced has the function in normalize.js rather than gatsby.node.js, should I move it there?

I'd like to do the following still:

- [ ] Progress Bar in CLI like I've seen during build elsewhere
- [ ] Configuration option to enable / disable cache? (or is this a feature that shouldn't have a toggle option?) 

I'll pull the git repo and get a PR going with what I have so far :)
